presto-main/src/main/java/com/facebook/presto/execution/QueryLimitMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,151 @@
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import static com.facebook.presto.execution.MemoryRevokingUtils.getMemoryPools;
+import static java.util.Objects.requireNonNull;
+
+public class QueryLimitMemoryRevokingScheduler
+{
+    private static final Logger log = Logger.get(QueryLimitMemoryRevokingScheduler.class);
+
+    private final ScheduledExecutorService taskManagementExecutor;
+    private final Function<QueryId, QueryContext> queryContextSupplier;
+
+    private final List<MemoryPool> memoryPools;
+    private final MemoryPoolListener memoryPoolListener = this::onMemoryReserved;
+
+    @Inject
+    public QueryLimitMemoryRevokingScheduler(
+            LocalMemoryManager localMemoryManager,
+            SqlTaskManager sqlTaskManager,
+            TaskManagementExecutor taskManagementExecutor)
+    {
+        this(
+                ImmutableList.copyOf(getMemoryPools(localMemoryManager)),
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getQueryContext,
+                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor());
+        log.debug("Using QueryLimitMemoryRevokingScheduler spilling strategy");
+    }
+
+    @VisibleForTesting
+    QueryLimitMemoryRevokingScheduler(
+            List<MemoryPool> memoryPools,
+            Function<QueryId, QueryContext> queryContextSupplier,
+            ScheduledExecutorService taskManagementExecutor)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.queryContextSupplier = requireNonNull(queryContextSupplier, "queryContextSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerPoolListeners();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool, QueryId queryId, long queryTotalMemoryUse)
+    {
+        try {
+            QueryContext queryContext = queryContextSupplier.apply(queryId);
+            long maxTotalMemory = queryContext.getMaxTotalMemory();
+            if (queryTotalMemoryUse <= maxTotalMemory) {
+                return;
+            }
+
+            log.debug("Scheduling check for %s", queryId);
+            scheduleRevoking(queryContext, queryTotalMemoryUse, maxTotalMemory);
```

Comment:
How likely are we to repeatedly revoke the same query because when we schedule revoking we don't check if we already attempted revoking?

And filtering out so we have less of those is also pretty sketchy because it's a lost wakeup scenario if we rely on having revocation take place to make progress.

The whole arrangement is actually a bit sketchy when you don't have a periodic check to unblock things.

The way this arrangement normally works safely is that when we schedule revocation we first do the revocation, then check if revocation is necessary/would be beneficial and go do it.

The reason you have the loop when processing the notification is to make sure that while you processed the notification you didn't accidentally lose/drop any notifications. That's generally a pretty safe arrangement when trying to reduce/filter them out.

Hope this actually make sense.

Reply:
yeah, just spoke about this with @pettyjamesm and decided to remove the checkPending setting. It wasn't working as intended, and if you're allocating more memory, it's not unlikely that you do need to revoke more.  However, you'll probably over-revoke since the queryTotalMemoryUse won't be correct (doesn't take into account memory that's been revoked in the meantime).

I could change to make it allow only one query at a time, and then deal with the issue of possible not revoking when you need to.

Possible ways to deal with not revoking memory when you should if I do change it.
1.  Adding the periodic check like the other spill strategies have.  The other memory revoking strategies use this, but it always seemed pretty hacky to me (and we'd have to add calls to get the memory reservation for each query, so that seems strictly worse than option 2) 
2. When you do revoke for a query, do it on a loop as you suggested until there is nothing else to be done. The downside is that we'd have to call for the queryTotalMemory usage from the memory pool in the loop during revokeMemory(). We just got rid of that to avoid extra locking.
3. wait during memory reservation if there's memory revocation scheduling in progress for the query. we'd have to be super careful about how the locking would work though.

What do you think?

Reply:
Re # 1
I think you don't need to it to cover for lost wakeups.
Re # 2
The loop will usually only have to run once, and only when revocation needs to occur. It's relatively rare so I would be concerned about the extra locking.

I think @pettyjamesm  was concerned about what happens in the non-revocation required case not the revocation required case. 99% of the time it's non-revocation and we want to avoid acquiring the lock, but the other 1% (or 0.1%) it's probably fine.

My concern about extra revocation is that it means we will create many small files for something like aggregation.

Reply:
> I think @pettyjamesm was concerned about what happens in the non-revocation required case not the revocation required case. 99% of the time it's non-revocation and we want to avoid acquiring the lock, but the other 1% (or 0.1%) it's probably fine.

Correct. I think a looping revoke / check for more revoking strategy that makes sure that no more than one revoke per query is in flight at once is probably not too big of a deal (famous last words).

> My concern about extra revocation is that it means we will create many small files for something like aggregation.

That hadn't occurred to me, but does sound like a potential cause for concern. That seems to be unavoidable regardless of whether the revokes are concurrent or looping per query though, right?

Reply:
as per discussion, updated to recheck the total memory reservation before starting revoking, so we don't do extra revocation. No need for a loop.